x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/ranges/ranges.tsx

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { i18n } from '@kbn/i18n';
+import { EuiForm } from '@elastic/eui';
+
+import {
+  autoInterval,
+  RangeType,
+  UI_SETTINGS,
+} from '../../../../../../../../src/plugins/data/common';
+import { RangeEditor } from './range_editor';
+import { OperationDefinition } from '../index';
+import { FieldBasedIndexPatternColumn } from '../column_types';
+import { updateColumnParam, changeColumn } from '../../../state_helpers';
+
+export const DEFAULT_INTERVAL = 1000;
+
+export const MODES = {
+  Range: 'range',
+  Histogram: 'histogram',
+} as const;
+
+export type RangeTypeLens = RangeType & { label: string };
+
+export type MODES_TYPES = typeof MODES.Range | typeof MODES.Histogram;
+
+export interface RangeIndexPatternColumn extends FieldBasedIndexPatternColumn {
+  operationType: 'range';
+  params: {
+    type: MODES_TYPES;
+    interval: 'auto' | number;
+    maxBars: 'auto' | number;
+    ranges: RangeTypeLens[];
+  };
+}
+
+export type RangeColumnParams = RangeIndexPatternColumn['params'];
+export type UpdateParamsFnType = <K extends keyof RangeColumnParams>(
+  paramName: K,
+  value: RangeColumnParams[K]
+) => void;
+
+export const isValidNumber = (value: number | '') =>
+  value !== '' && !isNaN(value) && isFinite(value);
+export const isRangeWithin = (range: RangeTypeLens): boolean => range.from <= range.to;
+const isFullRange = ({ from, to }: RangeType) => isValidNumber(from) && isValidNumber(to);
+export const isValidRange = (range: RangeTypeLens): boolean => {
+  const { from, to } = range;
+  return (isFullRange(range) && isRangeWithin(range)) || isValidNumber(from) || isValidNumber(to);
+};
+
+function getEsAggsParams({ sourceField, params }: RangeIndexPatternColumn) {
+  if (params.type === MODES.Range) {
+    return {
+      field: sourceField,
+      keyed: true,
```

Comment:
This property `keyed: true` does not relate to any of the properties in esaggs, so it's ignored.

Reply:
I've added because Visualize uses it when in Range mode:

<img width="740" alt="Screenshot 2020-09-10 at 12 46 50" src="https://user-images.githubusercontent.com/924948/92719716-dfc89d80-f363-11ea-957e-bc60433ef35e.png">

I've removed it here now, but it is hardcoded in Visualize for some reason.

Reply:
Looks like this is set by the agg config framework itself:
https://github.com/elastic/kibana/blob/5467f3164034784f10e60c084dbcc856ec0b4b36/src/plugins/data/common/search/aggs/buckets/range.ts#L106

It's a setting for how the response it structured:
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-range-aggregation.html#_keyed_response_7

But it looks like you don't need to set it to the esaggs config from the outside.